src/System.Data.SqlClient/src/System/Data/SqlClient/SqlTransaction.cs

```diff
@@ -141,47 +160,94 @@ override public void Commit()
 
         protected override void Dispose(bool disposing)
         {
-            if (disposing)
+            Exception e = null;
+            Guid operationId = s_diagnosticListener.WriteTransactionDisposeBefore(_isolationLevel, _connection);
+
+            try
+            {
+                if (disposing)
+                {
+                    if (!IsZombied && !IsYukonPartialZombie)
+                    {
+                        _internalTransaction.Dispose();
+                    }
+                }
+                base.Dispose(disposing);
+            }
+            catch (Exception ex)
+            {
+                e = ex;
+                throw;
+            }
+            finally
             {
-                if (!IsZombied && !IsYukonPartialZombie)
+                if (e != null)
                 {
-                    _internalTransaction.Dispose();
+                    s_diagnosticListener.WriteTransactionDisposeError(operationId, _isolationLevel, _connection, e);
+                }
+                else
+                {
+                    s_diagnosticListener.WriteTransactionDisposeAfter(operationId, _isolationLevel, _connection);
                 }
             }
-            base.Dispose(disposing);
         }
 
         override public void Rollback()
         {
-            if (IsYukonPartialZombie)
-            {
-                // Put something in the trace in case a customer has an issue
-                _internalTransaction = null; // yukon zombification
-            }
-            else
-            {
-                ZombieCheck();
+            Exception e = null;
+            Guid operationId = s_diagnosticListener.WriteTransactionRollbackBefore(_isolationLevel, _connection, null);
 
-                SqlStatistics statistics = null;
-                try
+            try
```

Comment:
@saurabh500 When `IsYukonPartialZombie == true`, I think we want to trace this case also, no?